rfcs/0000-compiled-runtime.md

```diff
@@ -0,0 +1,533 @@
+
+        return <div className={className} style={style} />;
+      }}
+    </ClassNames>
+  );
+}
+
+// 👇 transforms to
+
+import { i, ClassNames } from '@compiled/react/insert';
+
+function Comp(props) {
+  return (
+    <ClassNames>
+      {({ css }) => {
+        const [className, style] = css(
+          i(['_syaz4rde', ['._syaz4rde{color: var(--_kmurgp)}'], [['--_kmurgp', props.color]]])
+        );
+
+        return <div className={className} style={style} />;
+      }}
+    </ClassNames>
+  );
+}
+```
+
+And then style extract again is pointing to the extract runtime and stripping the style sheets.
+
+```diff
+-import { i, ClassNames } from '@compiled/react/insert';
++import { ClassNames } from '@compiled/react/extract';
+
+function Comp(props) {
+  return (
+    <ClassNames>
+      {({ css }) => {
+        const [className, style] = css(
+-          i(['_syaz4rde', ['._syaz4rde{color: var(--_kmurgp)}'], [['--_kmurgp', props.color]]])
++          ['_syaz4rde', undefined, [['--_kmurgp', props.color]]
+        );
+
+        return <div className={className} style={style} />;
+      }}
+    </ClassNames>
+  );
+}
+```
+
+## Drawbacks
+
+- More runtime bundle size would be added to the React package (however it might be a net reduction overall as every call site now uses an abstraction instead of the same markup)
+- Keyframes API doesn't fit naturally within this architecture
+- Dynamic declarations aren't supported in the ClassNames API without changing the API for consumers
+
+## Alternatives
+
+The primary alternative to discuss would be what data structure would get the best:
+
+- Runtime performance
+- Minification characteristics
+
+It does not need to be human readable as its a compilation target.
+
+## Adoption strategy
+
+This change would be a big shift from what how Compiled works today. The one consideration for adoption strategy will be ensuring the style extraction (strip runtime) is backwards compatible.
+
+To turn the new functionality on it should be introduced under some configuration where consumers can opt into the new compilation method which would then use what has been defined in this RFC.
+
+For Webpack this could look like:
+
+```js
+module.exports = {
+  module: {
+    rules: [
+      {
+        loader: '@compiled/webpack-loader',
+        options: {
+          experimental: {
+            localTransforms: true,
+          },
+        },
+      },
+    ],
+  },
+};
+```
+
+For the strip runtime it needs to be able to work with both transformation targets to ensure all sourced of Compiled transformed components will be extracted, with configuration to turn on/off legacy/experimental extraction.
+
+## Unresolved questions
+
+- What edge cases could arise with style extraction?
+- What performance characteristics (bundle size vs. runtime) are affected by this change?
+- This doesn't completely get rid of the need of module traversal (for style declaration static evaluation) - is it possible to?
```

Comment:
Why do you say this? I think it can, if you make assumptions about some usages (or guarantee it through types). The only other problem is stylesheet extraction, but it can leverage module traversal itself or de-optimise the output if necessary (leave anything imported un-extracted).

Reply:
Only for variable import use. The answer could be just throw and don't allow it? Problem solved. It's a similar aspect to Tom's point around completely changing APIs, for example if we were to get rid of the "0 config SSR" aspect would could in essence boil down the API to just.. css prop/class name. And styled for the short term conversion. With the obvious assumption the production if anyone wants SSR to work, well, they'd turn on config to extract anyway?

🤔

Reply:
Here's a spike https://github.com/atlassian-labs/compiled/compare/rfc-compiled-runtime-spike?expand=1

**Setup**

```
git checkout rfc-compiled-runtime-spike
yarn clean
yarn
```

**Tests**

```
yarn test packages/react/src/__tests__/experimental.test.tsx
yarn test packages/react/src/experimental.tsx
```

**Storybook**

```
yarn start
```

Removing the constraint of zero config (streaming) SSR really simplifies everything.

Reply:
Yeah, sounds good! I think I'd want a slightly different API in the end, but that's pretty easy to change.